internal/layers/layers_wcow.go

```diff
@@ -0,0 +1,601 @@
+//go:build windows
+// +build windows
+
+package layers
```

Comment:
This looks really good overall. I do have a few things I wanted to note:

- `ParseWCOWLayers` has to worry about a lot more than just parsing, since it returns a layers type that has to know how to mount itself.
- Because of the above point, `ParseWCOWLayers` also has to take aruments that should be unrelated to layer parsing. For instance, it seems weird that it would take a UVM argument. Also, if we add other parameters that are needed by some types of layers in the future, `ParseWCOWLayers` will have to keep growing with more arguments. This adds complexity and blurs abstractions.

Based on the above, I have a design change suggestion. It is still overall basically the same code you have here, just organized a little differently.

Rather than having separate types like `wcowHostForkedCIMLayers`, `wcowHostWCIFSLayers`, and `wcowIsolatedLayers` which each implement `WCOWLayers` (and thus provide `Mount`), instead provide interfaces like:

```go
type WCOWLayers interface{
	// Does nothing, just used as a marker for which types implement one of
	// the layer interfaces below.
	IsWCOWLayers()
}

type WCIFSLayers interface{
	WCOWLayers
	ScratchPath() string
	LayerPaths() []string
}

type CimFSLayers interface{
	WCOWLayers
	ScratchPath() string
	Layers() []CimFSLayer
}

type CimFSLayer struct{
	LayerPath string
	CimPath string
}
```

Then `ParseWCOWLayers` would return a type that either implements `WCIFSLayers` or that implements `CimFSLayers`<sup>[1](#1)</sup>. This also means we could easily have different types based on where we get the layer info. e.g. one `WCIFSLayers` impl for `layerFolders []string`, and one for `rootfs []*types.Mount`.

We would then have functions in `layers` package like the following:

```go
func MountWCOWHostLayers(ctx context.Context, layers WCOWLayers, containerID string) (*MountedWCOWLayers, resources.ResourceCloser, error) {
	...
}

func MountWCOWUVMLayers(ctx context.context, layers WCOWLayers, containerID string, uvm *uvm.UtilityVM) (*MountedWCOWLayers, resources.ResourceCloser, error) {
	...
}
```

These functions would do a type switch on `layers` to see if it is a `WCIFSLayers` or `CimFSLayers`, and act accordingly. If they are passed a type that they can't support (e.g. `MountWCOWUVMLayers` is passed a `CimFSLayers`), they would return an error.

In addition to addressing the concerns at the top, this design also keeps the code that knows how to mount layers<sup>[2](#2)</sup> separated from the code that parses layers. The parsing just has to know how to parse layer info into a common piece of information, which is much simpler.

<a name="1">[1]</a> If the interfaces of `WCIFSLayers` and `CimFSLayers` are too similar (e.g. interface ambiguity), they could instead provide `GetWCIFSLayers`/`GetCimFSLayers` functions that return structs.

<a name="2">[2]</a> As part of the other refactoring I'm doing, I'm considering introducing types like `WCOWHostSandbox` and `WCOWUVMSandbox`, which would manage the environment the containers run in. In this case the `Mount*` functions above could become part of these types, which would be convenient.

Reply:
Updated the PR as per suggestions. Let me know what you think.